anonymizer/OpenDiffix.Service/CSVDataProvider.fs

```diff
@@ -18,22 +18,123 @@ let private openCsvStream stream =
 
   csv
 
-type DataProvider(stream: TextReader) =
+type private TypeInfo = { IsEmpty: bool; IsBoolean: bool; IsInteger: bool; IsReal: bool }
+
+let private unknownTypeInfo = { IsEmpty = true; IsBoolean = true; IsInteger = true; IsReal = true }
+let private integerTypeInfo = { IsEmpty = false; IsBoolean = false; IsInteger = true; IsReal = true }
+
+let private parseBoolean (field: string) =
+  if field = "1" then
+    Boolean true
+  else if field = "0" then
+    Boolean false
+  else
+    match Boolean.TryParse field with
+    | (true, value) -> Boolean value
+    | (false, _) -> Null
+
+let private parseInt64 (field: string) =
+  match Int64.TryParse field with
+  | (true, value) -> Integer value
+  | (false, _) -> Null
+
+let private parseDouble (field: string) =
+  match Double.TryParse field with
+  | (true, value) -> Real value
+  | (false, _) -> Null
+
+let private tryInferType (field: string) =
+  if field = "" then
+    Some unknownTypeInfo
+  else
+    Some
+      {
+        IsEmpty = false
+        IsBoolean = parseBoolean field <> Null
+        IsInteger = parseInt64 field <> Null
+        IsReal = parseDouble field <> Null
+      }
+
+let private deriveColumnType (columnTypeInfo: TypeInfo) =
+  if columnTypeInfo.IsEmpty then StringType
+  else if columnTypeInfo.IsBoolean then BooleanType
+  else if columnTypeInfo.IsInteger then IntegerType
+  else if columnTypeInfo.IsReal then RealType
+  else StringType
+
+let private inferColumnTypes columnIndices (typeInfosByRow: TypeInfo option array list) =
+  let inferred =
+    columnIndices
+    |> List.map (fun i ->
+      typeInfosByRow
+      |> Seq.map (fun rowTypeInfo -> rowTypeInfo.[i].Value)
+      |> Seq.fold
+        (fun (acc: TypeInfo) fieldTypeInfo ->
+          { acc with
+              IsEmpty = acc.IsEmpty && fieldTypeInfo.IsEmpty
+              IsBoolean = acc.IsBoolean && fieldTypeInfo.IsBoolean
+              IsInteger = acc.IsInteger && fieldTypeInfo.IsInteger
+              IsReal = acc.IsReal && fieldTypeInfo.IsReal
+          }
+        )
+        unknownTypeInfo
+    )
+
+  let row = Array.create (List.max columnIndices + 1) (UnknownType "")
+
+  for inferredIndex, columnIndex in List.indexed columnIndices do
+    row.[columnIndex] <- deriveColumnType inferred.[inferredIndex]
+
+
+  row
+
+let private parse columnType field =
+  match columnType with
+  | BooleanType -> parseBoolean field
+  | IntegerType -> parseInt64 field
+  | RealType -> parseDouble field
+  | _ -> String field
+
+type DataProvider(stream: TextReader, streamForTypes: TextReader) =
   let csv = openCsvStream stream
+  let csvForTypes = openCsvStream streamForTypes
```

Comment:
Is there a better way to "reset" the reader. We need to scan the file twice - first on the construction (1000 rows max, for type inference) and second time on `OpenTable` which yields rows. Keeping two `StreamReader` instances was the only thing I managed to find :/.

Reply:
Yup, this would work. The question is though: how do you think, is it worth it? It's hard for me to judge without plunging in, how involved would the cache code be in practice.